include/libopencm3/stm32/f4/dma2d.h

```diff
@@ -0,0 +1,222 @@
+
+/* DMA2D Control Register */
+#define DMA2D_CR			MMIO32(DMA2D_BASE + 0x0U)
+#define DMA2D_CR_MODE_SHIFT		16
+#define DMA2D_CR_MODE_MASK		0x3
+#define DMA2D_CR_MODE_M2M		0	/* memory to memory */
+#define DMA2D_CR_MODE_M2MWPFC		1	/* memory to memory with pix convert */
+#define DMA2D_CR_MODE_M2MWB		2	/* memory to memory with blend */
+#define DMA2D_CR_MODE_R2M		3	/* register to memory */
+#define DMA2D_CR_CEIE			(1 << 13)
+#define DMA2D_CR_CTCIE			(1 << 12)
+#define DMA2D_CR_CAEIE			(1 << 11)
+#define DMA2D_CR_TWIE			(1 << 10)
+#define DMA2D_CR_TCIE			(1 << 9)
+#define DMA2D_CR_TEIE			(1 << 8)
+#define DMA2D_CR_ABORT			(1 << 2)
+#define DMA2D_CR_SUSP			(1 << 1)
+#define DMA2D_CR_START			(1 << 0)
+
+/* DMA2D Interrupt Status Register */
+#define DMA2D_ISR			MMIO32(DMA2D_BASE + 0x4U)
+#define DMA2D_ISR_CEIF			(1 << 5)
+#define DMA2D_ISR_CTCIF			(1 << 4)
+#define DMA2D_ISR_CAEIF			(1 << 3)
+#define DMA2D_ISR_TWIF			(1 << 2)
+#define DMA2D_ISR_TCIF			(1 << 1)
+#define DMA2D_ISR_TEIF			(1 << 0)
+
+/* DMA2D Interrupt Flag Clear Register */
+#define DMA2D_IFCR			MMIO32(DMA2D_BASE + 0x8U)
+#define DMA2D_IFCR_CCEIF		(1 << 5)
+#define DMA2D_IFCR_CCTCIF		(1 << 4)
+/* this is what the reference manual specifies */
+#define DMA2D_IFCR_CAECIF		(1 << 3)
+/* this is what it "should be" */
+#define DMA2D_IFCR_CCAEIF		(1 << 3)
+#define DMA2D_IFCR_CTWIF		(1 << 2)
+#define DMA2D_IFCR_CTCIF		(1 << 1)
+#define DMA2D_IFCR_CTEIF		(1 << 0)
+
+/* DMA2D Foreground Memory Address Register */
+#define DMA2D_FGMAR			MMIO32(DMA2D_BASE + 0xCU)
+
+/* DMA2D Foreground Offset Register */
+#define DMA2D_FGOR			MMIO32(DMA2D_BASE + 0x10U)
+#define DMA2D_FGOR_LO_SHIFT		0
+#define DMA2D_FGOR_LO_MASK		0x3fff
+
+/* DMA2D Background Memory Address Register */
+#define DMA2D_BGMAR			MMIO32(DMA2D_BASE + 0x14U)
+
+/* DMA2D Background Offset Register */
+#define DMA2D_BGOR			MMIO32(DMA2D_BASE + 0x18U)
+#define DMA2D_BGOR_LO_SHIFT		0
+#define DMA2D_BGOR_LO_MASK		0x3fff
+
+/* DMA2D Foreground PFC Control Register */
+#define DMA2D_FGPFCCR			MMIO32(DMA2D_BASE + 0x1cU)
+#define DMA2D_FGPFCCR_ALPHA_SHIFT	24
+#define DMA2D_FGPFCCR_ALPHA_MASK	0xff
+#define DMA2D_FGPFCCR_AM_SHIFT		16
+#define DMA2D_FGPFCCR_AM_MASK		0x3
+#define DMA2D_FGPFCCR_AM_NONE		0
+#define DMA2D_FGPFCCR_AM_FORCE		1
+#define DMA2D_FGPFCCR_AM_PRODUCT	2
+#define DMA2D_FGPFCCR_CS_SHIFT		8
+#define DMA2D_FGPFCCR_CS_MASK		0xff
+#define DMA2D_FGPFCCR_START		(1 << 5)
+#define DMA2D_FGPFCCR_CCM_ARGB8888	(0 << 4)
+#define DMA2D_FGPFCCR_CCM_RGB888	(1 << 4)
+#define DMA2D_FGPFCCR_CM_SHIFT		0
+#define DMA2D_FGPFCCR_CM_MASK		0xf
+#define DMA2D_FGPFCCR_CM_ARGB8888	0
+#define DMA2D_FGPFCCR_CM_RGB888		1
+#define DMA2D_FGPFCCR_CM_RGB565		2
+#define DMA2D_FGPFCCR_CM_ARGB1555	3
+#define DMA2D_FGPFCCR_CM_ARGB4444	4
+#define DMA2D_FGPFCCR_CM_L8		5
+#define DMA2D_FGPFCCR_CM_AL44		6
+#define DMA2D_FGPFCCR_CM_AL88		7
+#define DMA2D_FGPFCCR_CM_L4		8
+#define DMA2D_FGPFCCR_CM_A8		9
+#define DMA2D_FGPFCCR_CM_A4		10
```

Comment:
Personally, I prefer one macro. With two macros, after some copy pasting you risk to use the "wrong" one anyway.